### ISSUE TITLE
empty group cause error

### ISSUE THREAD

**Issue by HendryLeo** [zorn96/ms_active_directory#59]
If we try to find members from an empty AD group, error is returned

>>> session.find_members_of_group('python-test2',attributes_to_lookup=desired_attrs)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/odoo15/odoo15-venv/lib/python3.10/site-packages/ms_active_directory/core/ad_session.py", line 2131, in find_members_of_group
    single_map = self.find_members_of_groups([group], attributes_to_lookup, controls,
  File "/opt/odoo15/odoo15-venv/lib/python3.10/site-packages/ms_active_directory/core/ad_session.py", line 2178, in find_members_of_groups
    all_member_dns_set.update(member_dns)
TypeError: 'NoneType' object is not iterable

![image](https://user-images.githubusercontent.com/10069724/230856931-b434d83c-1021-4804-a2f1-ebcfe2d1becd.png)

**Comment by HendryLeo** [comment]
i changed the following function in ad_session.py

```
    def find_members_of_groups(self, groups: List[Union[str, ADGroup]], attributes_to_lookup: List[str] = None,
                               controls: List[Control] = None,
                               skip_validation: bool = False) -> Dict[Union[str, ADGroup],
                                                                      List[Union[ADUser, ADGroup,
                                                                                 ADComputer, ADObject]]]:
        """ Find the members of one or more groups in the domain, along with attributes of the members.

        :param groups: A list of either strings or ADGroups to look up the members of.
        :param attributes_to_lookup: Attributes to look up about the members of each group.
        :param controls: A list of LDAP controls to use when performing the search. These can be used to specify
                         whether or not certain properties/attributes are critical, which influences whether a search
                         may succeed or fail based on their availability.
        :param skip_validation: If true, assume all members exist and do not raise an error if we fail to look one up.
                                Instead, a placeholder object will be used for members that could not be found.
                                Defaults to False.
        :return: A dictionary mapping groups from the input list to lists of objects representing their members.
                 The objects may be of type ADUser, ADComputer, ADGroup, etc. - this function attempts to cast all
                 member objects to the most accurate object type representing them. ADObject will be used for members
                 that do not match any of the more specific object types in the library
                 (e.g. foreign security principals).
        :raises: InvalidLdapParameterException if any groups are not strings or ADGroups
        :raises: ObjectNotFoundException if any groups cannot be found.
        :raises: DomainSearchException if skip_validation is False and any group members cannot be found.
        """
        all_member_dns_set = set()
        group_to_member_dn_map = {}
        for group in groups:
            if isinstance(group, ADGroup):
                # if our group is an ADGroup, we can do the lookup by DN which is most efficient
                informed_group_object = self.find_group_by_distinguished_name(group.distinguished_name,
                                                                              [ldap_constants.AD_ATTRIBUTE_MEMBER],
                                                                              controls=controls)
            elif isinstance(group, str):
                informed_group_object = self.find_group_by_name(group, [ldap_constants.AD_ATTRIBUTE_MEMBER],
                                                                controls=controls)
            else:
                raise InvalidLdapParameterException('All groups must be either strings or ADGroup objects. {} is not.'.format(group))

            if informed_group_object is None:
                raise ObjectNotFoundException('No group could be found in the domain with the Group object class using group {}'.format(group))

            # add our members to our set and update our map
            if informed_group_object.get('member') is not None:
                member_dns = informed_group_object.get(ldap_constants.AD_ATTRIBUTE_MEMBER)
                all_member_dns_set.update(member_dns)
                group_to_member_dn_map[group] = member_dns

        logger.debug('Found %s unique members across %s groups', len(all_member_dns_set), len(groups))
        # now we need to look up all of the members.
        member_dn_to_objects = {}
        for member_dn in all_member_dns_set:
            ad_obj = self.find_object_by_distinguished_name(member_dn, attributes_to_lookup, controls=controls)
            if ad_obj is None:
                if skip_validation:
                    # use a placeholder object with just the dn
                    ad_obj = ADObject(member_dn, {}, self.domain)
                else:
                    raise DomainSearchException('An object with distinguished name {} was listed as a member of a '
                                                'group but information about it could not be found when searching the '
                                                'domain. This may indicate that the object was deleted in parallel '
                                                'with this search being performed, or an issue with the permissions '
                                                'of this session. If you believe it was a race with the domain '
                                                'changing, please retry this operation. Otherwise, please examine the '
                                                'permissions of the session.'.format(member_dn))
            member_dn_to_objects[member_dn] = ad_obj
        group_to_members_map = {}
        for group in groups:
            group_to_members_map[group] = []
            try:
                group_member_dns = group_to_member_dn_map[group]
                for dn in group_member_dns:
                    member_obj = member_dn_to_objects[dn]
                    group_to_members_map[group].append(member_obj)
            except:
                pass
        return group_to_members_map

```

**Comment by Pinkbyte** [comment]
The same solution as in previous comment, but in diff file:

```Python
--- a/ad_session.py     2021-12-09 21:10:33.000000000 +0300
+++ b/ad_session.py     2023-09-28 13:14:18.000000000 +0300
@@ -2174,9 +2174,11 @@
                 raise ObjectNotFoundException('No group could be found in the domain with the Group object class using '
                                               'group {}'.format(group))
             # add our members to our set and update our map
-            member_dns = informed_group_object.get(ldap_constants.AD_ATTRIBUTE_MEMBER)
-            all_member_dns_set.update(member_dns)
-            group_to_member_dn_map[group] = member_dns
+            # add check for group with empty members
+            if informed_group_object.get('member') is not None:
+                member_dns = informed_group_object.get(ldap_constants.AD_ATTRIBUTE_MEMBER)
+                all_member_dns_set.update(member_dns)
+                group_to_member_dn_map[group] = member_dns
 
         logger.debug('Found %s unique members across %s groups', len(all_member_dns_set), len(groups))
         # now we need to look up all of the members.
@@ -2199,11 +2201,15 @@
 
         group_to_members_map = {}
         for group in groups:
-            group_member_dns = group_to_member_dn_map[group]
+            # add check for group with empty members
             group_to_members_map[group] = []
-            for dn in group_member_dns:
-                member_obj = member_dn_to_objects[dn]
-                group_to_members_map[group].append(member_obj)
+            try:
+                group_member_dns = group_to_member_dn_map[group]
+                for dn in group_member_dns:
+                    member_obj = member_dn_to_objects[dn]
+                    group_to_members_map[group].append(member_obj)
+            except:
+                pass
         return group_to_members_map
 
     def find_members_of_group_recursive(self, group: Union[str, ADGroup], attributes_to_lookup: List[str] = None,
```

**Comment by zorn96** [comment]
hello hello - I was away for a bit because I got super busy but I'm back and I've resolved to be less of a deadbeat about this library.
@Pinkbyte @HendryLeo - thanks for finding and developing fixes for this issue! if y'all would like the credit for it in the library, feel free to put out a pull request for this and I'll happily merge it. otherwise, I can do so. I'm looking to have a new release out in the new year (some businesses use this so I won't release a new version in december in case people are doing a bad job of pinning their dependencies)
